File: bench/logparse/logparse.js
// Log line parsing: string building, trimming, splitting, prefix tests and
// integer conversion, with the results counted in a map.
const lines = [];
let i = 0;
while (i < 4000) {
    lines.push("  level=info user=u" + String(i % 97) + " ms=" + String(i % 31) + "  ");
    i = i + 1;
}

const users = new Map();
let elapsed = 0;
let round = 0;
while (round < 280) {
    let j = 0;
    while (j < lines.length) {
        const fields = lines[j].trim().split(" ");
        let k = 0;
        while (k < fields.length) {
            const field = fields[k];
            if (field.startsWith("user=")) {
                const user = field.slice("user=".length);
                users.set(user, (users.get(user) ?? 0) + 1);
            }
            if (field.startsWith("ms=")) {
                elapsed = elapsed + parseInt(field.slice("ms=".length), 10);
            }
            k = k + 1;
        }
        j = j + 1;
    }
    round = round + 1;
}
console.log(users.size, elapsed);

File: bench/logparse/logparse.py
# Log line parsing: string building, trimming, splitting, prefix tests and
# integer conversion, with the results counted in a dict.
lines = []
i = 0
while i < 4000:
    lines.append("  level=info user=u" + str(i % 97) + " ms=" + str(i % 31) + "  ")
    i = i + 1

users = {}
elapsed = 0
round_ = 0
while round_ < 280:
    j = 0
    while j < len(lines):
        fields = lines[j].strip().split(" ")
        k = 0
        while k < len(fields):
            field = fields[k]
            if field.startswith("user="):
                user = field[len("user="):]
                users[user] = users.get(user, 0) + 1
            if field.startswith("ms="):
                elapsed = elapsed + int(field[len("ms="):])
            k = k + 1
        j = j + 1
    round_ = round_ + 1
print(len(users), elapsed)
